### PR TITLE
S3

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3_client.c
Comment (graebm) [inline source/s3_client.c:65]: I think you need to move the
```
client->shutdown_called = true;
bool destroy_after_shutdown = client->release_called;
```
lines down here, after you've called `Py_CLEAR(client->on_shutdown)`

Imagine this scenario:
- python object is still alive
- s_s3_client_shutdown() runs
- sets shutdown_called = true
- invokes on_shutdown
  - from on_shutdown callback, user drops reference to python object
  - now capsule destructor runs, s_s3_client_release runs, sees that shutdown_called==true and release_called==true, so destroys memory
- on_shutdown callback returns, you run `Py_CLEAR(client->on_shutdown);` but crash because client memory has been deleted

I fixed this exact same kind of bug a month ago so it's fresh in my mind

File: source/s3_client.c
Comment (graebm) [inline source/s3_client.c:22]: trivial: this fn is only ever called from the capsule_destructor. why not just combine the functions?

File: source/s3_client.c
Comment (graebm) [inline source/s3_client.c:167]: Ugh. setup & shutdown error-handling code is always so fragile because you need to account for setup failing halfway through. For these reasons, I've come to prefer creating the capsule as soon as possible, rather than creating it as the last step. That way, if anything goes wrong I can just decref the capsule and let its destructor clean things up, even if setup only partially succeeded. Personally, I've found it simpler to write shutdown code that can handle a partially set up thing, instead of setup code that can shut down a partially setup thing, in addition to regular shutdown code.

So the order of fallible setup operations would be like:
```
s3_client = aws_mem_calloc()
capsule = PyCapsule_New(...)
if (!capsule)
    aws_mem_destroy(s3_client)
    return NULL

s3_client->on_shutdown = on_shutdown_py
Py_INCREF(on_shutdown)

s3_client->native = aws_s3_client_new()
if (!native)
    PyErr_SetAwsLastError()
    goto error;
return capsule
error:
Py_DECREF(capsule);
return NULL
```

but cleanup needs to check for the case that the native S3 client never actually existed
```
--- s_capsule_destructor() ---
aws_s3_client_release(client->native) // yeah, may be NULL, but release() can handle it
client->release_called=true;
s_destroy_if_ready(client)

--- s_on_shutdown() ---
fire callbacks
Py_CLEAR(client->on_shutdown)
client->shutdown_called = true
s_destroy_if_ready(client)

--- s_destroy_if_ready() ---
if (client->native && (!client->shutdown_called || !client->release_called))
    // native client successfully created, but not ready to clean up yet
    return

Py_XDECREF(client->on_shutdown) // just in case native never existed and shutdown never happened
aws_mem_release(client)
```

File: awscrt/s3.py
Comment (graebm) [inline awscrt/s3.py:198]: trivial: In HTTP, we set a similar future's result to be the `status_code`
could do the same here

File: awscrt/s3.py
Comment (graebm) [reply awscrt/s3.py:156]: also document the `shutdown_event` attribute

File: test/test_s3.py
Comment (graebm) [inline test/test_s3.py:62]: trivial: class constants are usually all caps like `REGION = "us-west-2"`

File: test/test_s3.py
Comment (graebm) [inline test/test_s3.py:68]: instance attributes are usually set in an init fn like:
```
def __init__(self):
    super().__init__()
    self.response_headers = None
    ...
```

File: test/test_s3.py
Comment (graebm) [inline test/test_s3.py:94]: having an assert here isn't going to do anything, these assert functions just raise exceptions
But this is on a callback from C on another thread
and as I mentioned earlier, exceptions on these callbacks don't do anything to stop the operation. PyErr_WriteUnraisable() will write something to stderr, but everything is going to keep charging along

the pattern I've been using for detecting errors in void callbacks is to do like:
```
    if chunk is None:
        self.callback_error = "the body chunk is none"
```

then at the end of the test on the main thread, after all the callbacks have fired, I check if any of them reported an error, like: `self.assertIsNone(self.callback_error)`

File: test/test_s3.py
Comment (graebm) [inline test/test_s3.py:121]: nice job btw
I saw how carefully you're setting the `shutdown_event` without affecting the refcount of anything else

File: source/s3_client.c
Comment (TingDaoK) [reply source/s3_client.c:110]: Oh, so the python object is not none, `aws_py_get_credentials_provider` should have a non-null returned value?

File: awscrt/s3.py
Comment (TingDaoK) [reply awscrt/s3.py:198]: feels like we can just set no result since the invalid response status code will result in a failure from native client.

File: awscrt/s3.py
Comment (TingDaoK) [reply awscrt/s3.py:71]: seems like we are doing the same thing for mqtt. So, if we set the default to None, we need to treat it as an object in the C land?

File: source/s3_meta_request.c
Comment (TingDaoK) [inline source/s3_meta_request.c:185]: self_py has been cleared here

File: awscrt/auth.py
Comment (JonathanHenson) [inline awscrt/auth.py:259]: I don't love this name. Maybe "new_custom" or something? I dunno. @graebm is the naming hero around her.

File: awscrt/auth.py
Comment (JonathanHenson) [inline awscrt/auth.py:270]: how will this work with something like ECDSA keys? @bretambrose how you feel about this one?

File: awscrt/s3.py
Comment (JonathanHenson) [inline awscrt/s3.py:41]: what's MR?

File: s3_benchmark.py
Comment (JonathanHenson) [inline s3_benchmark.py:75]: We probably should batch these before even calling into python in the first place.

File: s3_benchmark.py
Comment (JonathanHenson) [inline s3_benchmark.py:103]: can we bind out the new numa aware event-loop-groups here and give this a go on node 0?

File: s3_benchmark.py
Comment (JonathanHenson) [inline s3_benchmark.py:128]: isn't chunk NULL when using the FILE* api?

File: source/s3_meta_request.c
Comment (JonathanHenson) [inline source/s3_meta_request.c:84]: Do we have to have the GIL at this point? could we delay it until we're ready to perform the callback?

File: source/s3_meta_request.c
Comment (JonathanHenson) [inline source/s3_meta_request.c:129]: this will be slow as hell under heavy concurrency load.

File: s3_benchmark.py
Comment (TingDaoK) [reply s3_benchmark.py:128]: The on_body callback will not be invoked when the FILE pointer is set.

File: awscrt/s3.py
Comment (TingDaoK) [reply awscrt/s3.py:41]: Copy/pasted from https://github.com/awslabs/aws-c-s3/blob/main/include/aws/s3/s3_client.h#L54-L57.
Should stand for meta_request. Hmm, since we are not taking meta request to python land, we probably should get out of it🤔

File: source/s3_client.h
Comment (graebm) [inline source/s3_client.h:1]: trivial: maybe just name this file s3.h instead of s3_client.h
It's simpler to just have 1 header per submodule

File: source/s3_client.h
Comment (graebm) [inline source/s3_client.h:10]: trivial: don't need to include anything but module.h in this header file

File: source/s3_client.c
Comment (graebm) [inline source/s3_client.c:29]: trivial: just delete the "in case native never existed" comment. this code path is also taken if native DID exist and normal shutdown is complete. The comment above this, about "not ready to clean up yet" conveys this all pretty well

File: source/s3_client.c
Comment (graebm) [inline source/s3_client.c:44]: 🤣
"C land"
I love it

**Review by graebm — COMMENTED** [review]
all trivial feedback. Going to wait until this is merged with the other PRs before reviewing further

**Review by graebm — COMMENTED** [review]
still reading, but here's some more feedback

File: source/s3_meta_request.c
Comment (TingDaoK) [reply source/s3_meta_request.c:74]: `s_s3_meta_request_capsule_destructor` cleaned those up. So, if anything fails, the capsule destructor should still be invoked, right?

File: source/s3_meta_request.c
Comment (graebm) [inline source/s3_meta_request.c]: you're might remove the http_message_py variable from existence anyway...

but if you keep it, this isn't the right place to clear it
it still needs to exist until the shutdown callback fires, right?
because if we cancel other threads might still be reading from it?

File: source/s3_meta_request.c
Comment (graebm) [inline source/s3_meta_request.c]: is it possible for `destroy_after_shutdown` to ever be false?

if it's not possible, let's not even have this variable

File: source/s3_meta_request.c
Comment (graebm) [inline source/s3_meta_request.c]: shouldn't we just be calling `s_destroy_if_ready()`?
having cleanup code in many different places is ... error prone as time goes on and more people touch this code

File: source/s3_meta_request.c
Comment (graebm) [inline source/s3_meta_request.c:366]: return AWS_OP_ERR;

File: source/s3_meta_request.c
Comment (graebm) [inline source/s3_meta_request.c:371]: you can't `return`! We're holding the GIL!

File: source/s3_meta_request.c
Comment (graebm) [reply source/s3_meta_request.c:71]: yeah, I saw that later. cool

File: source/s3_meta_request.c
Comment (TingDaoK) [reply source/s3_meta_request.c:228]: eh... I am confused by myself🤯. Probably related to some crazy lifetime management before. Will remove!

File: source/s3_meta_request.c
Comment (TingDaoK) [reply source/s3_meta_request.c]: I think it's probably better to have either `None` or `bytes`? Since the headers can be `None` as well. Will document it!

File: source/s3_meta_request.c
Comment (TingDaoK) [reply source/s3_meta_request.c]: Yep, removed!

File: source/s3_meta_request.c
Comment (graebm) [inline source/s3_meta_request.c:132]: 🤣

**Review by graebm — APPROVED** [review]
beautiful!

**Pull request merged**
